feat(utxo-lib): various improvements to utxo-lib and related libs

OttoAllmendinger opened the pull request:
See commits for details

Issue: BTC-0


<!--
# Please be aware of the following when making your pull request:

## Description

Please include a summary of your proposed changes and which issue is being addressed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Issue Number

Internal Users - Please include the related internal tracking number (e.g. BG-000000).
External Users - Please link to any relevant github issues as necessary.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] My code compiles correctly for both Node and Browser environments
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My commits follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) and I have properly described any BREAKING CHANGES
- [ ] The ticket or github issue was included in the commit message as a reference
- [ ] I have made corresponding changes to the documentation and on any new/updated functions and/or methods - [jsdoc](https://jsdoc.app/)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
-->

OttoAllmendinger committed:
feat(utxo-lib): add extractTransaction to UtxoPsbt

Improve type signature of extractTransaction to return UtxoTransaction to avoid
having to cast the return value.

Issue: BTC-0

OttoAllmendinger committed:
feat(blockapis): support bitgo.TxOutPoint in fetchInputs

More flexibility.

Issue: BTC-0

OttoAllmendinger committed:
feat(utxo-lib): improve type signatures for `createTransactionFrom*`

The new signatures make sure that the passing `amountType` results in a narrower
return type. The result of
`createTransactionFromBuffer(tx, network, { amountType: 'bigint' })` is now
`UtxoTransaction<bigint>` instead of `UtxoTransaction<number | bigint>`.

Mark `createTransactionFromHex` as deprecated. It is only a wrapper around
`createTransactionFromBuffer` and we need to repeat the extra type declarations.
It also does not have the `version` parameter, making it inconsistent with
`createTransactionFromBuffer`. It is better to handle the decoding at the call
site.

Issue: BTC-0

OttoAllmendinger committed:
feat(utxo-lib): export PsbtInput, PsbtOutput from index.ts

Deprecated PsbtInputType alias

Issue: BTC-0

OttoAllmendinger committed:
fix(utxo-lib): use deterministic keys in testutil/mock.ts

Deterministic tests rock

Issue: BTC-0

OttoAllmendinger committed:
feat(utxo-bin): add option `--hex`

For smaller transactions it is fine to pass them as an argument

Issue: BTC-0

mmcshinsky-bitgo commented on modules/utxo-lib/src/bitgo/index.ts:
```diff
@@ -1,3 +1,5 @@
+export { PsbtInput, PsbtOutput } from 'bip174/src/lib/interfaces';
```
```suggestion
export { PsbtInput, PsbtOutput } from 'bip174/;
```
Can we get to these without going into `src`?

OttoAllmendinger replied:
unfortunately not (hence the re-export here)

OttoAllmendinger committed:
feat(utxo-lib): add, use equalPublicKey in UtxoPsbt

Under certain conditions we might be passing a 33-byte plain pubkey where a
32-byte pubkey is expected or vice versa in cases where we don't actually care
about the first byte. Using `equalPublicKey` helps simplify the comparisons
where we only want to know of the x coordinate is equal.

Issue: BTC-0

OttoAllmendinger committed:
refactor(utxo-lib): rename equalPublicKey to equalPublicKeyIgnoreY

Only use equalPublicKeyIgnoreY in taproot methods (there was one site that
wasn't taproot-related)

Issue: BTC-0

OttoAllmendinger committed:
fix(utxo-lib): add parameter `ignoreY` to deriveKeyPair

For taproot sites we compare the result ignoring the Y coordinate

Issue: BTC-0

OttoAllmendinger committed:
refactor(utxo-lib): add deriveKeyPairForInput

Add wrapper function to derive key pair for input.

Issue: BTC-0
